hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/SnapshotContext.java

```diff
@@ -134,66 +173,120 @@ boolean isTerminalSnapshot() {
-            activeSnapshotId = currentSnapshotId;
+            activeSnapshotIdPhase1 = currentSnapshotId;
         } else {
             // the snapshot will be started once all higher priority sources are done
             // see #taskletDone()
             logger.info("Snapshot " + snapshotId + " for " + jobNameAndExecutionId + " is postponed" +
                     " until all higher priority vertices are completed (number of such vertices = "
-                    + numHigherPriorityTasklets + ')');
+                    + numPrioritySsTasklets + ')');
         }
-        if (numTasklets == 0) {
+        if (numSsTasklets == 0) {
             // member is already done with the job and master didn't know it yet - we are immediately successful
             return completedFuture(new SnapshotOperationResult(0, 0, 0, null));
         }
-        future = new CompletableFuture<>();
-        return future;
+        phase1Future = new CompletableFuture<>();
+        return phase1Future;
+    }
+
+    /**
+     * This method is called when the member received {@link
+     * SnapshotCompleteOperation}.
+     */
+    synchronized CompletableFuture<Void> startNewSnapshotPhase2(long snapshotId, boolean success) {
+        if (snapshotId == activeSnapshotIdPhase2) {
+            // this is possible in case when the operation packet was lost and retried
+            logger.warning("Second request for phase 2 snapshot with id=" + snapshotId);
+            CompletableFuture<Void> res = this.phase2Future;
+            if (res == null) {
+                // if there's no ongoing phase 2, we don't know what was the response. We reply with
+                // 100% success. For regular snapshots the failures are ignored. They are reported to
+                // the user only for exported snapshot. We neglect that.
+                res = completedFuture(null);
+            }
+            return res;
+        }
+        assert snapshotId == activeSnapshotIdPhase1 : "requested phase 2 for snapshot ID " + snapshotId
+                + ", but phase 1 snapshot ID is " + activeSnapshotIdPhase1;
+        assert phase1Future == null : "phase 1 still ongoing";
+        assert phase2Future == null : "phase 2 already in progress";
+        assert snapshotId > activeSnapshotIdPhase2
+                : "new snapshotId for phase 2 not larger than previous. Previous=" + activeSnapshotIdPhase2 + ", new="
+                        + snapshotId;
+        assert numPTasklets >= 0 : "numPTasklets=" + numPTasklets;
+        if (isCancelled) {
+            throw new CancellationException("execution cancelled");
+        }
+        this.lastPhase1Successful = success;
+        assert numPrioritySsTasklets == 0 : "numPrioritySsTasklets=" + numPrioritySsTasklets;
+
+        boolean casSuccess = numRemainingTasklets.compareAndSet(0, numPTasklets);
+        assert casSuccess : "numRemainingTasklets wasn't 0, but " + numRemainingTasklets.get();
```

Comment:
this should be a failure, not just assertion

Reply:
There's no way the user can cause it. I hit it during development. It's a check for further changes that we expect will be met.